internal/concierge/apiserver/apiserver.go

```diff
@@ -107,39 +107,45 @@ func (c completedConfig) New() (*PinnipedServer, error) {
 		return nil, fmt.Errorf("could not install API groups: %w", err)
 	}
 
-	shutdown := &sync.WaitGroup{}
+	controllersShutdownWaitGroup := &sync.WaitGroup{}
+	controllersCtx, cancelControllerCtx := context.WithCancel(context.Background())
+
 	s.GenericAPIServer.AddPostStartHookOrDie("start-controllers",
 		func(postStartContext genericapiserver.PostStartHookContext) error {
 			plog.Debug("start-controllers post start hook starting")
+			defer plog.Debug("start-controllers post start hook completed")
 
-			ctx, cancel := context.WithCancel(context.Background())
-			go func() {
-				defer cancel()
-
-				<-postStartContext.StopCh
```

Comment:
This is where the deadlock was happening. This `StopCh` channel is not going to be closed until our pre-shutdown hook finishes running. But our pre-shutdown hook calls `shutdown.Wait()` which is effectively waiting for this goroutine to end (because this goroutine cancels the context which allows the `runControllers()` call to stop blocking, which in turn ends the `WaitGroup` that the pre-shutdown hook it waiting for).

Reply:
🤯